### PR TITLE
Enable setting up Kubernetes cluster in Ubuntu on Azure

### PR DESCRIPTION
**Pull request by WeixuZhuang**
Implement basic cloud provider functionality to deploy Kubernetes on
Azure.  SaltStack is used to deploy Kubernetes on top of Ubuntu 
virtual machines.  OpenVpn provides network connectivity.  For
kubelet authentication, we use basic authentication (username and 
password).  The scripts use the legacy Azure Service Management APIs.

We have set up a nightly test job in our Jenkins server for federated
testing to run the e2e test suite on Azure.  With the cloud provider
scripts in this commit, 14 e2e test cases pass in this environment.
We plan to implement additional Azure functionality to support more
test cases.

---

This change is [Reviewable](http://reviewable.k8s.io/reviews/kubernetes/kubernetes/21207)

### CODE REVIEW

**Comment by parantap2001** [comment]
@ixdy @fejta - thoughts ?

**Comment by ixdy** [comment]
We're in code slush for 1.2, and this is a pretty big change, so it may take a little while to get this reviewed and merged (and it's unlikely to make 1.2). I'll look around to see who'd be appropriate to review.

**Comment by WeixuZhuang** [comment]
@ixdy , I found that there are some merged in cluster/azure which is some code for setting up kubernetes in azure cloud. My code does the same thing and it has been here for two weeks waiting for review.... In addition, my code is completed and we are running nightly end-to-end test on Azure using my code. If you want to merge code for Azure set up, why you leave my code unreviewed here for weeks and merges someone else's code?

**Comment by ixdy** [comment]
@WeixuZhuang sorry for the delay - as I mentioned earlier, we were trying to stabilize the tree in preparation for cutting the `release-1.2` branch, and so now that that's done, there should be more traction on this PR. Note that you'll need to separate out the docs into a separate PR, as mentioned by the bot and in https://groups.google.com/forum/#!topic/kubernetes-dev/UedwIYb-fHg.

This PR also needs to be rebased. I'll look to see if there's someone who can review this after that's done.

**Comment by davidopp** [comment]
cc/ @colemickens

**Comment by parantap2001** [comment]
https://github.com/kubernetes/kubernetes/pull/22231/commits
here makes the entire cluster bring up on azure/e2e on kubernetes dependent on an external repository here - https://github.com/colemickens/azkube

@ixdy @colemickens @jeffmendoza @fejta @davidopp @WeixuZhuang

Why did we choose this approach rather than picking this existing approach similar to other clusters -
kubernetes as is here #21207 

We can continue with the rebase to get this code in, however before we do that we wanted consensus on the best direction to move forward in terms on cluster bring up.

FYI. We have several E2E running here -
https://console.cloud.google.com/storage/browser/kube_azure_log/
and can easily upstream results to be shared in the dashboard of the nightly runs here -
http://storage.googleapis.com/kubernetes-test-history/static/index.html

**Comment by davidopp** [comment]
Very sorry, I was not aware of this PR (and I suspect @colemickens was not either). Although it may not have helped, filing an issue, and then attaching the PR to an issue, makes it easier for us to track stuff, because it will at least get assigned to a team. With 333 open PRs right now, it's very easy for issue-less PRs to get lost.

**Comment by colemickens** [comment]
I was also unaware of that pull request.

I'm working on getting e2e validation for the new Azure bring-up.

I had a preliminary version that was fully shell-script based, but it was error-prone and it's behavior was hard to get correct depending on the version of the azure-xplat-cli that was installed. The method used in my pull request minimized the number of dependencies needed ahead-of-time and increased the likelihood of a successful deployment. It also gives us some flexibility - the core piece (producing an ARM template, creating the Azure Service Principal, etc) can be re-used for different setups. For example, `azkube` currently creates CoreOS templates for Azure, but could be extended in the future to produce templates for Ubuntu or others, as well.

Additionally, I moved away from the Salt stack bringup based on indications that a more self-contained, docker-based bring up was the way forward.

**Comment by parantap2001** [comment]
point taken @davidopp. Miss on our part, we should have created an issue for it. Completely understandable given the huge churn for v1.2 as well.

@colemickens We are playing with the patches that you have and can evaluate whether generating a ubuntu template with your containerized methodology is simpler. However I am not convinced on the fact that any changes to cluster/azure now potentially means pull requests to two repositories.
Additionally I am a fan of uniformity so would prefer the methodology adopted and supported for other clusters (aws, gce..)

**Comment by WeixuZhuang** [comment]
@ixdy rebased/removed docs.

@davidopp created issue to track status #22715

@colemickens created a shell switch to have two different ways of setting up the cluster/azure kube-up. the switch is 'local' to retain the shell script methodology and 'docker' to use the container method for cluster/azure setup.

**Comment by zmerlynn** [comment]
What's the status of this PR? I'm finally iterating my backlog and it's not clear if this PR is going to get acted on, or if we're headed a different way for Azure.

**Comment by parantap2001** [comment]
Changes in this PR are still current and rebase correctly on master.

The e2e tests on azure still work based on this patchset and are posted regularly here -
https://console.cloud.google.com/storage/browser/kube_azure_log/

We were hoping this change would go in as it does not break existing ways on deploying on azure (i.e externally maintained container with coreos + flannel introduced with a commit recently). In fact it adds a new way of deploying cluster/azure with ubuntu + openvpn.
This commit also provides a flag to switch between the method of deployment so both methodologies can co-exist.

**Comment by davidopp** [comment]
I'll reassign to myself for now.

**Comment by spiffxp** [comment]
/cc @kubernetes/sig-testing 

Would be in favor of seeing this eventually get exercised via a daily kubernetes-e2e-azure job, same as is currently done for AWS with [kubernetes-e2e-aws](https://console.cloud.google.com/storage/browser/kubernetes-jenkins/logs/kubernetes-e2e-aws/)

**Comment by colemickens** [comment]
I have a slight concern that this is using the legacy Azure Service Management APIs rather than the new Azure Resource Manager APIs. This has large implications down the line for the cloudprovider implementation as well. Additionally, the heavy use of Salt, just as `kube-deploy` is being started with the intention of moving the deployment stuff out-of-tree and as I understand it, away from Salt.

I'm in progress on getting a kubernetes-e2e-azure job working for the new Azure bring-up that now exists in the tree. I'm adding it in a similar way to how the kubernetes-e2e-aws job is defined in the JJB definitions in hack/jenkins.

**Comment by spiffxp** [comment]
>Additionally, the heavy use of Salt, just as kube-deploy is being started with the intention of moving the deployment stuff out-of-tree and as I understand it, away from Salt.

@colemickens I suspect that effort is going to take some time; if this more or less works today, it's something to at least iterate on, no?

**Comment by jeffmendoza** [comment]
If this is working well for AppFormix, could we merge it under `cluster/azure-legacy`?

**Comment by parantap2001** [comment]
We have been running an e2e test nightly for the past several weeks.  14 scenarios are passing. We are fine with moving this code to `cluster/azure-legacy`.  Please let us know if there is consensus to this approach before we make the change.

File: docs/getting-started-guides/README.md [hunk 1]
```diff
@@ -34,7 +34,6 @@ Documentation for other releases can be found at
 
 This file has moved to: http://kubernetes.github.io/docs/getting-started-guides/
 
-
```
Comment (bgrant0607) [inline docs/getting-started-guides/README.md]: Please revert this change.

**Comment by brendandburns** [comment]
Given a desire to move forward (and the sometimes slow pace of the new work).  I'd be in favor of merging this as azure-legacy as well.

@zmerlynn @ixdy @davidopp @colemickens any objections?

@parantap2001 how hard is it to update this to the more modern Azure Resource Manager APIs?

File: cluster/saltbase/salt/kube-master-addons/kube-master-addons.sh [hunk 2]
```diff
@@ -22,23 +22,23 @@ while true; do
 
   if which docker 1>/dev/null 2>&1; then
 
-    timeout 30 docker load -i /srv/salt/kube-bins/kube-apiserver.tar 1>/dev/null 2>&1
+    timeout 120 docker load -i /srv/salt/kube-bins/kube-apiserver.tar 1>/dev/null 2>&1
```
Comment (mikedanese) [inline cluster/saltbase/salt/kube-master-addons/kube-master-addons.sh]: cc @dchen1107

Comment (WeixuZhuang) [inline cluster/saltbase/salt/kube-master-addons/kube-master-addons.sh]: I increased the time for docker load to avoid unnecessary timeout when doing docker load. I have encounter an issue that some times 30 seconds is not enough for docker load

**Comment by parantap2001** [comment]
@brendandburns regarding the transition to Azure Resource Manager Apis. I believe all the azure_call <actions> used in `cluster/azure-legacy/util.sh` need to change to populate a json as described below. Followed by a bootstrap call with the json.

https://github.com/Azure/azure-content/blob/master/articles/xplat-cli-azure-resource-manager.md
https://github.com/Azure/azure-quickstart-templates/blob/master/101-vm-simple-linux/azuredeploy.json

We need to try it but seems simple. 

We can pick that up as the next set of commits on top of this as that seems to be supported way forward in azure.

**Comment by travisnewhouse** [comment]
Thanks to @rmmh, the Azure results are now available the kubernetes-test-history page:
http://storage.googleapis.com/kubernetes-test-history/static/index.html

In order to maintain these nightly tests, it will be helpful to have these deployment scripts merged.  @WeixuZhuang has addressed comments and rebased.

@zmerlynn @ixdy @davidopp @colemickens @brendandburns Are there any remaining concerns before this can be merged?

**Comment by WeixuZhuang** [comment]
Hi @brendandburns I have code using Azure Resource Manager API ready(which is based on this PR). I think we will also switch to use ARM for nightly test soon. Would you like me to push out the code? I prefer to wait until this code get merged and push out as a new PR since it is already lots of code here.

Thanks,

**Comment by travisnewhouse** [comment]
@zmerlynn @ixdy @davidopp @colemickens @brendandburns Friendly ping to bring this to your attention again.  Are there any remaining concerns before this can be merged?

File: cluster/saltbase/salt/docker/init.sls [hunk 3]
```diff
@@ -47,6 +47,93 @@ docker:
...
+       - pkgrepo: 'apt-key'
+
+docker:
+   service.running:
+     - enable: True
+     - require:
+       - file: {{ environment_file }}
+     - watch:
+       - file: {{ environment_file }}
+
```
Comment (davidopp) [inline cluster/saltbase/salt/docker/init.sls]: is there a missing `endif` here?

File: cluster/saltbase/salt/kube-master-addons/kube-master-addons.sh (hunk 2 above)
Comment (davidopp) [inline cluster/saltbase/salt/kube-master-addons/kube-master-addons.sh]: ping @dchen1107

File: cluster/saltbase/salt/top.sls [hunk 4]
```diff
@@ -58,6 +61,9 @@ base:
     - kube-controller-manager
     - kube-scheduler
     - supervisor
+{% if grains['cloud'] is defined and not grains.cloud in ['aws', 'gce', 'vagrant', 'vsphere'] %}
```
Comment (davidopp) [inline cluster/saltbase/salt/top.sls]: Why did you add this?

Comment (WeixuZhuang) [inline cluster/saltbase/salt/top.sls]: since we need nginx installed to bring up the cluster

File: cluster/saltbase/salt/docker/init.sls (hunk 3 above)
Comment (WeixuZhuang) [inline cluster/saltbase/salt/docker/init.sls]: the endif is in line 498,

elsif in line 50 is the else for if in line 10

File: cluster/saltbase/salt/top.sls (hunk 4 above)
Comment (davidopp) [inline cluster/saltbase/salt/top.sls]: Sorry, what I meant was, why isn't this
`{% if grains['cloud'] is defined and grains['cloud'] == 'azure-legacy' %}`
?

File: cluster/saltbase/salt/docker/init.sls (hunk 3 above)
Comment (davidopp) [inline cluster/saltbase/salt/docker/init.sls]: Oh sorry, I didn't see that the next line after this one is an `elif`. Never mind.

File: cluster/saltbase/salt/top.sls (hunk 4 above)
Comment (WeixuZhuang) [inline cluster/saltbase/salt/top.sls]: oh, it is my bad, I will push out the fix. Thanks

**Comment by davidopp** [comment]
ok to test

**Comment by davidopp** [comment]
LGTM

**Comment by davidopp** [comment]
Test failure is
```
Verifying ./hack/../hack/verify-boilerplate.sh
Boilerplate header is wrong for: /go/src/k8s.io/kubernetes/cluster/azure-legacy/templates/create-kubeconfig.sh
FAILED   ./hack/../hack/verify-boilerplate.sh	1s
```

**Comment by WeixuZhuang** [comment]
@davidopp I have fixed the issue with boilerplate check.  All tests are passing now.

Thanks

**Comment by Krylon360** [comment]
@WeixuZhuang would love to have the ARM templates added to this. ARM is pretty much the standard now with Azure. ASM still works; however many features are being depreciated very quickly. It's best to be ahead of the depreciation of ASM by having the ARM scripts ready to go.

**Comment by parantap2001** [comment]
@Krylon360 ARM Changes are ready. We are in the process of making a separate issue and adding commits to it.

For this commit/PR we are at a stable checkpoint, as the federated dashboard has these results upstreamed nightly now -
https://k8s-testgrid.appspot.com/k8s#azure

**Comment by jamescw** [comment]
@WeixuZhuang the submit queue seems blocked, how can I find the install instructions for this in the meantime?

**Comment by travisnewhouse** [comment]
@davidopp Rebased again.  Can you let us know if there are any other requirements before this can be merged?  Thanks!

**Comment by davidopp** [comment]
Hi Travis. I feel bad because it took us so long to review this, and then by the time it was finished we were in our "code freeze" and not accepting new PRs until we cut the 1.3 final release. Since this is a new feature and not fixing existing functionality, I think we should wait until the code freeze is over, which I believe will be around the end of this week. Sorry about the delay.

So in a nutshell, there's nothing blocking this other than the lifting of the code freeze.

**Comment by davidopp** [comment]
ok to test

**Comment by davidopp** [comment]
ok to test

File: cluster/get-kube.sh [hunk 5]
```diff
@@ -30,6 +30,8 @@
 #   * export KUBERNETES_PROVIDER=aws; wget -q -O - https://get.k8s.io | bash
 #  Libvirt (with CoreOS as a guest operating system)
 #   * export KUBERNETES_PROVIDER=libvirt-coreos; wget -q -O - https://get.k8s.io | bash
+#  Microsoft Azure
```
Comment (davidopp) [inline cluster/get-kube.sh]: @colemickens you might want to add a comment for yours to this file too

**Comment by davidopp** [comment]
LGTM
